### PR TITLE
[refactor](exchange) Optimize the logic related to sending and closing

### PR DESCRIPTION
**Pull request by mrhhsg**
## Proposed changes

Issue Number: close #xxx

### CODE REVIEW

**Comment by mrhhsg** [comment]
run buildall

File: be/src/vec/sink/vdata_stream_sender.cpp [hunk 1]
```diff
@@ -166,131 +166,51 @@ Status PipChannel::send_broadcast_block(std::shared_ptr<BroadcastPBlockHolder>&
...
-        return send_local_block(exec_status, eos);
-    }
-    if (eos) {
-        RETURN_IF_ERROR(_serializer.serialize_block(_ch_cur_pb_block, 1));
+        return _send_local_block(exec_status, eos);
     }
-    RETURN_IF_ERROR(send_remote_block(_ch_cur_pb_block, eos, exec_status));
-    ch_roll_pb_block();
+    RETURN_IF_ERROR(send_remote_block(_pblock.release(), eos, exec_status));
     return Status::OK();
```
Comment (yiguolei) [inline be/src/vec/sink/vdata_stream_sender.cpp:174]: useless code

File: be/src/vec/sink/vdata_stream_sender.cpp [hunk 2]
```diff
@@ -166,131 +166,51 @@ Status PipChannel::send_broadcast_block(std::shared_ptr<BroadcastPBlockHolder>&
...
-            _local_recvr->remove_sender(_parent->sender_id(), _be_number, exec_status);
-        }
-        return Status::OK();
-    } else {
-        _serializer.reset_block();
-        return _receiver_status;
+    if (!block.empty() || eos) {
+        RETURN_IF_ERROR(send_local_block(&block, eos, true));
     }
+    _serializer.reset_block();
```
Comment (yiguolei) [inline be/src/vec/sink/vdata_stream_sender.cpp:186]: 如果这里调用reset，那么181 行的意义是什么？

File: be/src/vec/sink/vdata_stream_sender.cpp [hunk 3]
```diff
@@ -166,131 +166,51 @@ Status PipChannel::send_broadcast_block(std::shared_ptr<BroadcastPBlockHolder>&
...
-    }
-    return Status::OK();
-}
 
-template <typename Parent>
-Status Channel<Parent>::add_rows(Block* block, const std::vector<uint32_t>& rows, bool eos) {
-    if (_fragment_instance_id.lo == -1) {
+        if (eos) [[unlikely]] {
+            _local_recvr->remove_sender(sender_id, _be_number, Status::OK());
+            LOG(INFO) << "****** remove sender: " << sender_id << ", be_number: " << _be_number;
```
Comment (yiguolei) [inline be/src/vec/sink/vdata_stream_sender.cpp:207]: too many logs

File: be/src/vec/sink/vdata_stream_sender.cpp [hunk 4]
```diff
@@ -166,131 +166,51 @@ Status PipChannel::send_broadcast_block(std::shared_ptr<BroadcastPBlockHolder>&
...
     if (is_local()) {
-        return send_local_block(exec_status, eos);
-    }
-    if (eos) {
-        RETURN_IF_ERROR(_serializer.serialize_block(_ch_cur_pb_block, 1));
+        return _send_local_block(exec_status, eos);
     }
-    RETURN_IF_ERROR(send_remote_block(_ch_cur_pb_block, eos, exec_status));
-    ch_roll_pb_block();
+    RETURN_IF_ERROR(send_remote_block(_pblock.release(), eos, exec_status));
```
Comment (yiguolei) [inline be/src/vec/sink/vdata_stream_sender.cpp:173]: 这么写有风险，假如有异常，罗指针没释放

File: be/src/vec/sink/vdata_stream_sender.cpp [hunk 5]
```diff
@@ -166,131 +166,51 @@ Status PipChannel::send_broadcast_block(std::shared_ptr<BroadcastPBlockHolder>&
...
-    }
-    RETURN_IF_ERROR(send_remote_block(_pblock.release(), eos, exec_status));
-    return Status::OK();
-}
-
-template <typename Parent>
-Status Channel<Parent>::send_current_block(bool eos, Status exec_status) {
-    // FIXME: Now, local exchange will cause the performance problem is in a multi-threaded scenario
-    // so this feature is turned off here by default. We need to re-examine this logic
+Status PipChannel::_send_current_block(bool eos, Status exec_status) {
```
Comment (yiguolei) [inline be/src/vec/sink/vdata_stream_sender.cpp:169]: 把pipchannel和channel 合并了吧，代码特别乱

**Comment by mrhhsg** [comment]
run buildall

File: be/src/vec/sink/vdata_stream_sender.cpp [hunk 6]
```diff
@@ -166,135 +159,48 @@ Status PipChannel::send_broadcast_block(std::shared_ptr<BroadcastPBlockHolder>&
...
-    if (serialized) {
-        RETURN_IF_ERROR(send_current_block(false, Status::OK()));
-    }
-
-    return Status::OK();
 }
 
-template <typename Parent>
-Status Channel<Parent>::close_wait(RuntimeState* state) {
+Status Channel::close_wait(RuntimeState* state) {
```
Comment (yiguolei) [inline be/src/vec/sink/vdata_stream_sender.cpp:203]: 这个close wait 函数还有意义吗？

File: be/src/pipeline/exec/exchange_sink_operator.h [hunk 7]
```diff
@@ -112,8 +112,8 @@ class ExchangeSinkLocalState final : public PipelineXSinkLocalState<> {
         return Status::OK();
     }
     Status _send_new_partition_batch();
-    std::vector<vectorized::PipChannel*> channels;
-    std::vector<std::shared_ptr<vectorized::PipChannel>> channel_shared_ptrs;
+    std::vector<vectorized::Channel*> channels;
```
Comment (yiguolei) [inline be/src/pipeline/exec/exchange_sink_operator.h:115]: 为什么保留2个？

File: be/src/vec/sink/vdata_stream_sender.cpp [hunk 8]
```diff
@@ -309,67 +215,52 @@ Status Channel<Parent>::close_wait(RuntimeState* state) {
...
-        return status;
+        RETURN_IF_ERROR(send_remote_block(nullptr, true, Status::OK()));
     }
+    return Status::OK();
 }
 
-template <typename Parent>
-Status Channel<Parent>::close(RuntimeState* state, Status exec_status) {
-    if (_closed) {
+Status Channel::_wait_last_brpc() {
```
Comment (yiguolei) [inline be/src/vec/sink/vdata_stream_sender.cpp:238]: 我们都走exchange sink buffer了，为啥还需要wait rpc，这个操作应该是在exchange sink buffer 里完成的

File: be/src/vec/sink/vdata_stream_sender.h [hunk 9]
```diff
@@ -193,47 +211,23 @@ class Channel {
...
     const TUniqueId _fragment_instance_id;
     PlanNodeId _dest_node_id;
 
     // the number of RowBatch.data bytes sent successfully
     int64_t _num_data_bytes_sent {};
     int64_t _packet_seq {};
 
-    bool _need_close;
-    bool _closed;
+    bool _need_close {};
```
Comment (yiguolei) [inline be/src/vec/sink/vdata_stream_sender.h:225]: 明确写上，是true还是false

**Comment by mrhhsg** [comment]
run buildall

**Comment by mrhhsg** [comment]
run buildall

File: be/src/vec/sink/vdata_stream_sender.cpp [hunk 10]
```diff
@@ -166,210 +151,76 @@ Status PipChannel::send_broadcast_block(std::shared_ptr<BroadcastPBlockHolder>&
...
-    return Status::OK();
+    return send_remote_block(std::move(_pblock), eos, exec_status);
 }
 
-template <typename Parent>
-Status Channel<Parent>::send_local_block(Status exec_status, bool eos) {
+Status Channel::_send_local_block(Status exec_status, bool eos) {
     SCOPED_TIMER(_parent->local_send_timer());
     Block block = _serializer.get_block()->to_block();
     _serializer.get_block()->set_mutable_columns(block.clone_empty_columns());
```
Comment (yiguolei) [inline be/src/vec/sink/vdata_stream_sender.cpp:164]: 为什么要clone 一把

File: be/src/vec/sink/vdata_stream_sender.cpp [hunk 11]
```diff
@@ -147,13 +132,13 @@ Status PipChannel::send_remote_block(PBlock* block, bool eos, Status exec_status
...
     }
     return Status::OK();
 }
 
-Status PipChannel::send_broadcast_block(std::shared_ptr<BroadcastPBlockHolder>& block, bool eos) {
-    COUNTER_UPDATE(Channel<pipeline::ExchangeSinkLocalState>::_parent->blocks_sent_counter(), 1);
+Status Channel::send_broadcast_block(std::shared_ptr<BroadcastPBlockHolder>& block, bool eos) {
+    COUNTER_UPDATE(_parent->blocks_sent_counter(), 1);
     if (eos) {
         if (_eos_send) {
```
Comment (yiguolei) [inline be/src/vec/sink/vdata_stream_sender.cpp:143]: eos_send 是什么意思？

Comment (mrhhsg) [reply be/src/vec/sink/vdata_stream_sender.cpp:143]: Here use `_eos_send` to avoid sending eos repeatedly.

**Comment by mrhhsg** [comment]
run buildall

File: be/src/vec/sink/vdata_stream_sender.h [hunk 12]
```diff
@@ -177,151 +140,40 @@ class Channel {
...
 
-    Status add_rows(Block* block, const std::vector<uint32_t>& rows, bool eos) override {
-        if (Channel<pipeline::ExchangeSinkLocalState>::_fragment_instance_id.lo == -1) {
+    Status add_rows(Block* block, const std::vector<uint32_t>& rows, bool eos) {
+        if (_fragment_instance_id.lo == -1) {
             return Status::OK();
         }
 
         bool serialized = false;
         _pblock = std::make_unique<PBlock>();
```
Comment (Gabriel39) [inline be/src/vec/sink/vdata_stream_sender.h:164]: DCHECK(_pblock == nullptr)

**Comment by mrhhsg** [comment]
run buildall

**Comment by mrhhsg** [comment]
run buildall

**Comment by mrhhsg** [comment]
run buildall

**Review by yiguolei — DISMISSED** [review]

File: be/src/pipeline/exec/exchange_sink_operator.cpp [hunk 13]
```diff
@@ -106,6 +111,24 @@ Status ExchangeSinkLocalState::init(RuntimeState* state, LocalSinkStateInfo& inf
...
+    if (_finished_channels.contains(channel_id)) {
+        LOG(WARNING) << "query: " << print_id(_state->query_id())
+                     << ", on_channel_finished on already finished channel: " << channel_id;
+        return;
+    } else {
+        _finished_channels.emplace(channel_id);
+        if (_working_channels_count.fetch_sub(1) == 1) {
+            set_reach_limit();
+            if (_finish_dependency) {
+                _finish_dependency->set_ready();
```
Comment (Gabriel39) [inline be/src/pipeline/exec/exchange_sink_operator.cpp:126]: `_finish_dependency` should never block if only local channels remain

**Comment by mrhhsg** [comment]
run buildall

**Comment by mrhhsg** [comment]
run buildall

File: be/src/vec/sink/vdata_stream_sender.h [hunk 14]
```diff
@@ -313,12 +182,54 @@ class PipChannel final : public Channel<pipeline::ExchangeSinkLocalState> {
 
     std::shared_ptr<pipeline::Dependency> get_local_channel_dependency();
 
-private:
+protected:
+    Status _send_local_block(bool eos);
+    Status _send_current_block(bool eos);
+
+    bool _recvr_is_valid() {
```
Comment (HappenLee) [inline be/src/vec/sink/vdata_stream_sender.h:189]: should we return Status in this func and dispose the `_receiver_status` out the function is better ?

**Comment by mrhhsg** [comment]
run buildall

File: be/src/pipeline/exec/exchange_sink_buffer.cpp [hunk 15]
```diff
@@ -470,18 +424,13 @@ bool ExchangeSinkBuffer::_is_receiver_eof(InstanceLoId id) {
     return _instance_to_receiver_eof[id];
 }
 
-void ExchangeSinkBuffer::_turn_off_channel(InstanceLoId id, bool cleanup) {
+void ExchangeSinkBuffer::_turn_off_channel(InstanceLoId id) {
```
Comment (HappenLee) [inline be/src/pipeline/exec/exchange_sink_buffer.cpp:427]: if we call the function, we must get the lock of `std::unique_lock<std::mutex> lock(*_instance_to_package_queue_mutex[id]);` so better we can read the important msg from comment or function name

Comment (yiguolei) [inline be/src/pipeline/exec/exchange_sink_buffer.cpp:427]: Just add the lock as a reference parameter, then we could make sure we locked.

**Comment by mrhhsg** [comment]
run buildall

**Comment by mrhhsg** [comment]
run buildall

**Review by yiguolei — APPROVED** [review]
LGTM

**Review by Gabriel39 — APPROVED** [review]

**Comment by Gabriel39** [comment]
Need to summary this PR in description

**Pull request merged**